pkg/kapp/crdupgradesafety/validator_test.go

```diff
@@ -174,3 +175,104 @@ func TestNoStoredVersionRemoved(t *testing.T) {
 		})
 	}
 }
+
+func TestNoExistingFieldRemoved(t *testing.T) {
+	for _, tc := range []struct {
+		name        string
+		new         apiextensionsv1.CustomResourceDefinition
+		old         apiextensionsv1.CustomResourceDefinition
+		shouldError bool
+	}{
+		{
+			name: "no existing field removed, no error",
+			old: apiextensionsv1.CustomResourceDefinition{
+				Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+					Versions: []apiextensionsv1.CustomResourceDefinitionVersion{
+						{
+							Name: "v1alpha1",
+							Schema: &apiextensionsv1.CustomResourceValidation{
+								OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+									Type: "object",
+									Properties: map[string]apiextensionsv1.JSONSchemaProps{
+										"fieldOne": {
+											Type: "string",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			new: apiextensionsv1.CustomResourceDefinition{
+				Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+					Versions: []apiextensionsv1.CustomResourceDefinitionVersion{
+						{
+							Name: "v1alpha1",
+							Schema: &apiextensionsv1.CustomResourceValidation{
+								OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+									Type: "object",
+									Properties: map[string]apiextensionsv1.JSONSchemaProps{
+										"fieldOne": {
+											Type: "string",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "existing field removed, error",
+			old: apiextensionsv1.CustomResourceDefinition{
+				Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+					Versions: []apiextensionsv1.CustomResourceDefinitionVersion{
+						{
+							Name: "v1alpha1",
+							Schema: &apiextensionsv1.CustomResourceValidation{
+								OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+									Type: "object",
+									Properties: map[string]apiextensionsv1.JSONSchemaProps{
+										"fieldOne": {
+											Type: "string",
+										},
+										"fieldTwo": {
+											Type: "string",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			new: apiextensionsv1.CustomResourceDefinition{
+				Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+					Versions: []apiextensionsv1.CustomResourceDefinitionVersion{
+						{
+							Name: "v1alpha1",
+							Schema: &apiextensionsv1.CustomResourceValidation{
+								OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+									Type: "object",
+									Properties: map[string]apiextensionsv1.JSONSchemaProps{
+										"fieldOne": {
+											Type: "string",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			shouldError: true,
+		},
```

Comment:
Could we also add a test that verifies that if a new version is added to the CRD with the field removed that it doesn't return an error? IIUC it should be OK to add an entirely new version to the CRD with a different set of fields

Reply:
That's a good idea, I've addd a test in my [latest commit](https://github.com/carvel-dev/kapp/pull/922/commits/51b64005ce47e75e822f08f1bda3e656320bf0c3) that verifies that if a new version is added to the CRD with the removed field, it wouldn't return an error.